src/transformers/models/jamba/modeling_jamba.py

```diff
@@ -0,0 +1,2127 @@
+
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+
+    # Ignore copy
+    @add_start_docstrings_to_model_forward(JAMBA_INPUTS_DOCSTRING)
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Union[List[torch.FloatTensor], HybridMambaAttentionDynamicCache]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        output_router_logits: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, MoeModelOutputWithPast]:
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_router_logits = (
+            output_router_logits if output_router_logits is not None else self.config.output_router_logits
+        )
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # retrieve input_ids and inputs_embeds
+        if input_ids is not None and inputs_embeds is not None:
+            raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
+        elif input_ids is not None:
+            batch_size, seq_length = input_ids.shape
+        elif inputs_embeds is not None:
+            batch_size, seq_length, _ = inputs_embeds.shape
+        else:
+            raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+        past_key_values_length = 0
+
+        if self.gradient_checkpointing and self.training:
+            if use_cache:
+                logger.warning_once(
+                    "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
+                )
+                use_cache = False
+
+        if use_cache:
+            if isinstance(past_key_values, Cache) and not isinstance(
+                past_key_values, HybridMambaAttentionDynamicCache
+            ):
+                past_key_values = HybridMambaAttentionDynamicCache.from_legacy_cache(past_key_values.to_legacy_cache())
+            use_legacy_cache = not isinstance(past_key_values, HybridMambaAttentionDynamicCache)
+            if use_legacy_cache:
+                past_key_values = HybridMambaAttentionDynamicCache.from_legacy_cache(past_key_values)
+            past_key_values_length = past_key_values.get_usable_length(seq_length, self._attn_layer_index)
```

Comment:
a lot of this can be simplified with the new `cache_positions` used in `GemmaModel` for example

Reply:
1. I see `cache_positions` is not used in the `DynamicCache` which `HybridMambaAttentionDynamicCache` inherits from. So not sure how easy the integration will be
2. Getting this `HybridMambaAttentionDynamicCache` right was a bit of a struggle and it works now as expected. I think it's best that we make the move to `cache_positions` in a different PR and not let it block merging this PR. It shouldn't change any weight names so that won't be an issue.

Reply:
Yes no worries!

Reply:
actually it's not used, but `generate` uses it, and should be easy to implement! Now it's still up to you to have it in a follow up PR, but will simplify the code